pkg/policy/l4.go

```diff
@@ -963,12 +967,21 @@ func (l4 *L4Policy) AccumulateMapChanges(adds, deletes []identity.NumericIdentit
 	proto := uint8(l4Filter.U8Proto)
 	derivedFrom := l4Filter.DerivedFromRules
 
+	// Must take a copy of 'users' as GetNamedPortsMap() will lock the Endpoint below and
+	// the Endpoint lock may not be taken while 'l4.mutex' is held.
 	l4.mutex.RLock()
-	for epPolicy := range l4.users {
+	users := make(map[*EndpointPolicy]struct{}, len(l4.users))
+	for user := range l4.users {
+		users[user] = struct{}{}
+	}
+	l4.mutex.RUnlock()
+
+	for epPolicy := range users {
 		// resolve named port
 		if port == 0 && l4Filter.PortName != "" {
 			var err error
-			port, err = epPolicy.NamedPortsMap.GetNamedPort(l4Filter.PortName, proto)
+			npMap := epPolicy.PolicyOwner.GetNamedPortsMap(direction == trafficdirection.Ingress)
```

Comment:
nit: performing comparisons to get the bool here to pass into `GetNamedPortsMap()` suggests that the callee should be taking `trafficdirection` directly (or otherwise why do we have a dedicated `trafficdirection` type?).

Reply:
Will have to clean this up in a follow-on PR. In this case, without wider refactoring, we'll have to use comparisons somewhere, as trafficdirection is not universally used yet.